cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -255,29 +277,69 @@ public void visitEnd() {
+                                                     nodeProcessor.getActions(), paramPositions));
+        }
+
+        /** Helper Methods **/
+
+        private void verifyEntityParts(List<EntityPartDetail> entityPartDetails, Type enforceOn,
+                                       Map<Integer, ParameterDetail> parameterAnnotationNode,
+                                       Map<String, Integer> paramAnnotation) {
+          Type entityPartType = getEntityPartType(parameterAnnotationNode, paramAnnotation, entityPartDetails.get(0));
+          // if the first entity part is not string then it should be of same type specified in enforce on
+          // TODO: Later change this to even work with parent type as we will support enforceOn parent type
+          if (!entityPartType.equals(Type.getType(String.class))) {
+            Preconditions.checkArgument(entityPartType.equals(enforceOn), "Found invalid entity type '%s' for " +
+                                          "enforceOn '%s' in annotation on '%s' method in '%s' class.",
+                                        entityPartType.getClassName(), enforceOn.getClassName(), methodName, className);
+            AuthEnforceUtil.getEntityIdPartsCount(enforceOn);
+          } else {
+            // Since the entity part/parts provided is of String type so validate that we have sufficient part for
+            // EntityId creation
+            // verify that all parts are of type string
+            for (EntityPartDetail entityPartDetail : entityPartDetails) {
+              entityPartType = getEntityPartType(parameterAnnotationNode, paramAnnotation, entityPartDetail);
+              Preconditions.checkArgument(entityPartType.equals(Type.getType(String.class)),
+                                          "Found part %s of type %s in a multiple part entity specification of " +
+                                            "AuthEnforce. Only String is supported in multiple parts.",
+                                          entityPartDetail.getEntityName(), entityPartType);
+            }
+            int requiredSize = AuthEnforceUtil.getEntityIdPartsCount(enforceOn);
+            Preconditions.checkArgument(requiredSize == entityPartDetails.size(), "Found %s entity parts in " +
+                                          "AuthEnforce annotation on method %s in class %s to do enforcement on %s " +
+                                          "which requires %s entity parts or an %s", entityPartDetails.size(),
+                                        methodName, className, enforceOn, requiredSize, enforceOn.getClassName());
+
+          }
+        }
+
+        private Type getEntityPartType(Map<Integer, ParameterDetail> parameterAnnotationNode,
+                                       Map<String, Integer> paramAnnotation, EntityPartDetail entityPartDetail) {
+          return entityPartDetail.isField() ? fieldDetails.get(entityPartDetail.getEntityName()) :
```

Comment:
Seems better to just store the type inside the `EntityPartDetail` when being constructed.